### ISSUE TITLE
How to configure eglot to use flycheck?

### ISSUE THREAD

**Issue by wyuenho** [joaotavora/eglot#42]
I think eglot is great, but the default integration with flymake is lacking quite a bit, I was wondering if you could give a few pointers for those of us who prefer flycheck, or better yet, integrate with flycheck in eglot?

**Comment by joaotavora** [comment]
I could consider this, but perhapa you would first care to elaborate on the "lacking quite a bit" bit?

**Comment by joaotavora** [comment]
> It's documentation is basically non-existent, 

Have you read the [manua](https://www.gnu.org/software/emacs/manual/html_node/flymake/Using-Flymake.html#Using-Flymake)l? What part of it do you find basically (or sophisticatedly) non-existent? 

> flymake-cc is sooo slow 

Flymake-cc is a backend in an unreleased version master of Emacs. 

* What does this have to do with Flymake itself "lacking" anything?
* Have you made a bug report about the conditions in which 20% mem usage and CPU time happen? Use the command `M-x report-emacs-bug` (and CC the bug report to me, if you can, too).
* Why are you using flymake-cc in Eglot? In other words, what is the connection between the purportedly malfunctioning backend and the limitations of Flymake in general or in its integration with Eglot?
* Which customization group is broken?  Again, what is the relation?

**Comment by joaotavora** [comment]
Incidentally, you can find in the manual a section titled [Backend functions](https://www.gnu.org/software/emacs/manual/html_node/flymake/Backend-functions.html#Backend-functions).  You'll find that Flymake backends follow a well-defined protocol to communicate with the Flymake UI.  You might refer the Flycheck developers to this protocol and convince them to create a Flycheck interface for the Flymake backends.  It makes much more sense to do it in this direction since Flymake is a part of Emacs, and Flycheck isn't.

**Comment by wyuenho** [comment]
The manual says: "This manual is for GNU Flymake (version 0.3, April 2004)", so no, I had not.

> What does this have to do with Flymake itself "lacking" anything?

It comes with Emacs, is it not a part of flymake?

> Have you made a bug report about the conditions in which 20% mem usage and CPU time happen? No, I've already spent too much time on the mailing list in the past month, but now you know. I'll send a full report when I have time.

> Why are you using flymake-cc in Eglot?

Good question. All I have installed is cquery on my system, Flymake just decided to use flymake-cc for some reason.

> In other words, what is the connection between the purportedly malfunctioning backend and the limitations of Flymake in general or in its integration with Eglot?

Pretty sure it's flymake, but there's no connection between the malfunctioning backend and limitations of flymake. I only cited it as one of the examples of flymake's inadequacy. One of the most annoying UI problems of flymake is when my point is over a diagnostic, the problem is my shown on the echo area and I have move my mouse.

> Which customization group is broken and how? Again, what is the relation?

flymake-cc on master. Just M-x customize-group RET flymake-cc and you'll see.

**Comment by joaotavora** [comment]
> The manual says: "This manual is for GNU Flymake (version 0.3, April 2004)", so no, I had not.

So you found the manual, read an outdated timestamp, clicked nowhere else, and declared here that was equivalent to no documentation?  BTW, you also missed `M-x info`, the "Commentary" section of the library, NEWS entries, docstrings of public symbols, etc. All regular places to go looking for documentation.

The timestamp problem is already fixed for 26.2.

> No, I've already spent too much time on the mailing list in the past month, but now you know. 

*What* do I know?

You don't you think making a bug report is time better spent than vaguely surfacing a problem that you can't describe minimally?  Are you counting the time that you making me invest in these incomplete bug reports?  Is it really so hard to `M-x profiler-start` during the purported slowness and copy the result of `M-x profiler-report` here or in a `M-x report-emacs-bug`?

> Good question. All I have installed is cquery on my system, Flymake just decided to use flymake-cc for some reason.

This is bug somewhere flymake-cc, Flymake or Eglot. I'll fix it. Flymake-cc should probably give way to Eglot or the other way around. Anyway, stuff on master doesn't always work like in the released version (where, as explained, this doesn't happen).

In the meantime, you can

```
(add-hook 'c-mode-hook 
  (lambda () (remove-hook 'flymake-diagnostic-functions 'flymake-cc t)))
```

To remove `flymake-cc` from the equation.

> One of the most annoying UI problems of flymake is when my point is over a diagnostic, the problem is not shown in the echo area and I have move my mouse.

Have you tried `C-h .` or `M-x display-local-help`?  Or, if you like it automatically
```
(custom-set-variables
 '(help-at-pt-timer-delay 0.1)
 '(help-at-pt-display-when-idle '(flymake-diagnostic)))
```

This is issue #8, by the way and Emacs bug https://debbugs.gnu.org/cgi/bugreport.cgi?bug=31921

You can play with `help-at-pt-timer-delay` and `eldoc-idle-delay` to finely control who has the echo area if you happen to be sitting on both a Flymake diagnostic and eldoc-active area.

If this was "one of the most annoying UI problems", is it solved now? What are the others? 
  
> I only cited it as one of the examples of flymake's inadequacy. 

Excuse me, Jimmy, but how exactly is a problem that you are having in a specific backend that is in the unreleased development version of Emacs, and which you refuse to co-substantiate with actual information, an example of the "inedequacy" of Flymake in general or its integration in Eglot?

I really am asking honestly because you have (since the early days of Yasnippet many moons ago to very recently) usually provided me with useful bug reports and feature suggestions. This, on the other hand, really sounds like a gratuitous and unsubstantiated critique of the new Flymake from someone who would just like to keep using Flycheck because whatever (which is totally legitimate). If it's just the second part, I don't hold it against you in any way that and have already suggested a way to get Flycheck working in Eglot. If you really have specific, objective problems with Flymake (trivial as they might be, like colors or something like that) please describe them.

**Comment by wyuenho** [comment]
Look, I don't want to hurt your feelings or argue about Flymake on a thread purely about eglot, that's why I gave you minimum information.

I know you've spent a tremendous amount of time reviving an obsolete built-in package, I admire that a lot, and it does work way better than before when I can get it to work, but it's still nowhere near as good as Flycheck, and many of them not your fault. It's simply because the community has moved on 6 years ago, and I even longer. But if you want a list of problems of Flymake, here they are:

1. Very limited number of backends OTTB. If I'm going to install a number of third-party packages, why on earth would I want to use Flymake? If I'm going to go back to the legacy backend, why did I abandon it 10 years ago?
2. I use about a dozen different languages, markup or otherwise, on a daily basis. Flycheck has many dozens of checkers OTTB, and many dozens more on Melpa, multiple ones for different purposes for the same languages. Most of the languages I use have at least one Flycheck checkers, Flymake has how many?
3. How many of these Flymake backends actually work with the new Flymake?
4. How exactly do I configure Flymake to work with my language? Does the manual say anything about it? Does the NEWS file? Does the Changlog? Does any variable or function docstring? Is it wise to read half a day just to find out how to get Flymake doing one check for one file when Flycheck **Just Works** OOTB with no configuration necessary? Useless information = no information. Adding that to all the other outdated information on the internet, I conclude that as good as not having any documentation.
5. Why would I think of looking at `*-help-*` functions when all I want to see is an diagnostics from **Flymake** in the echo area?
6. I already have Flycheck on my system that works better, why do I have to turn on Flymake consuming valuable memory just for eglot?
7. The version number of the built-in version of Flymake on 26.1 is still 0.3, why should I care about it?

Given all these non-sense, it's a miracle that I know Emacs 26.1 has revamped Flymake. 

In multiple places, you have said that Flymake is likely to still be inferior to Flycheck, I assume you already know of these problems. There's no reason for Flymake to continue to exist other than purely political reasons.

And no, I don't want to help improve Flymake, so don't ask me to hang out on the mailing list or use Debian's horrible beast of a bug tracker hacked up in the 90s. I've unsubscribed from the mailing list years ago because I know it's a major time sink, and I've only resubscribed very recently out of neccessity.

**Comment by joaotavora** [comment]
> Look, I don't want to hurt your feelings

You aren't hurting my feelings in any way, Jimmy. But you are wasting my time.

> but it's still nowhere near as good as Flycheck,

What, in the name of anything that's sacred to you, are the reasons for this sentence? I've recently collaborated with the Flycheck developers on a side-by-side comparison and despite minor things (that you are not even close to mentioning), there is feature parity.

> if you want a list of problems of Flymake, here they are:

1,2,3 are the same, and the only true objection. But it makes no sense in this context, because you talked about Eglot integration and Flymake has a backend for that, where notably, Flycheck hasn't yet.  

Yes there are only a few backends provided currently with Flymake.  Things take time to build.  What are the backends that are missing for you? In so many letters you could have a typed  the names of languages you would like to see supported instead.

> How exactly do I configure Flymake to work with my language? 

[LIKE THIS](https://www.gnu.org/software/emacs/manual/html_node/flymake/Backend-functions.html#Backend-functions)

> Does the manual say anything about?

[YES](https://www.gnu.org/software/emacs/manual/html_node/flymake/Backend-functions.html#Backend-functions)

> Does any variable or function docstring?

YES. `C-h v flymake-diagnostic-functions RET`.

> Flycheck Just Works OOTB with no configuration necessary

Does it guess the configuration flags for any arbitrary C project OOTB? I'd be very surprised?

> Why would I think of looking at *-help-* functions when all I want to see is an diagnostics from Flymake in the echo area?

Because that's how Emacs works! Aren't you using Emacs??  If you didn't "think of looking", that's OK, *you ask first*. When someone solves your problem you can optionally say thanks and not *b\*itch* that it's counter-intuitive to go looking for stuff in Emacs.

> I already have Flycheck on my system that works better, why do I have to turn on 
Flymake consuming valuable memory just for eglot?

It doesn't add any more memory than turning on Flycheck on that same buffer wouldn't.  Or do have any shred of evidence to the contrary?

> The version number of the built-in version of Flymake on 26.1 is still 0.3, why should I care about it?

You are caring a LOT about it.  Enough to rant on.  What's the version number got to do with anything?  Did it bother you in any moment when you turned on Flymake? And aren't using master anyway?  You want to base a serious critique of something on the fact that I forgot to update a version number, **seriously**?

> Given all these nonsense, 

Oh nice, a true statement.

> e, so don't ask me to hang out on the mailing list or use Debian's horrible beast of a bug tracker hacked up in the 90s

Again, what gratuitous, useless waste of words.  What is the relevance here that you feel this about the bug tracker (it's rude, but that's up to you). Also, I'm not asking you to *hang out* with me or the people in emacs-devel, thanks very much.

It's very very simple: to report a bug send mail to bug-gnu-emacs@gnu.org. Or is email obsolete these days too?

You don't have to interact with anything, or even to reply, but if you do that you've already proven more useful than in this entire exchange.

**Comment by brotzeit** [comment]
@wyuenho I'm sure this conversation would have been more productive if you would have asked politely if there's any chance that eglot will work with flycheck some day. Or even better, offer your help. I admit that I also prefer flycheck, but if he wants to try a different way with flymake, why not. You can still use lsp-mode with flycheck.

**Comment by wyuenho** [comment]
@brotzeit Which part of my original request is impolite? The whole point of the issue is to ask for pointers to integrate with flycheck, if Joao can do it himself, that's better because usually by the time I can get to it, Joao is already done, but that's not necessary as you can see from the title of this issue. That's also a tacit help offer to ease his workload. I dare not asking so many thing from an author working for free. This issue is not about Flymake, it's about eglot.

This whole shitstorm started because @joaotavora  is being defensive when I suggested Flymake to be inferior. Is it not inferior? Other than having integration with eglot, which part of it is not either me-too or straightly inferior? I don't even want to pull out the comparison table on Flycheck's website, they bother me sure, but not enough to "rant" about. It's sufficiently said by the Flycheck authors.

And to Joao, I don't know what triggered this reaction TBH, but it's rather unwarranted. I said I'll provide a full report when I have time. I'm not obligated to provide you a full report immediately, just like you are not obligated to grant all of my requests. Besides, I don't know how many times I have to say this, but this is about eglot, not Flymake. It's a simple yes or no that you'll provide the pointers. I can dig into it when I have time, so it's fine if you don't want to offer any help, as I don't see any willingness so far.

Also, a word of advice, I think it's about time the phrase "developer experience" enter emacs dev's vocabulary. When I see the version number on package's header and manual being the same as 14 year's ago, I'm not going to bother diffing the actual content of the files. No sensible person will do it and if I were to do this, my life would be unbearable due to a constant fact-checking and second-guessing information that's supposed to be reliable. Similarily, a declarative API is straightly better than screenfuls of documentation that I can't run as code, but this is a weaker objection.

About Flycheck's no configuration required bit, I've never had to configure Flycheck other than a few times where I had to add a mode to it's predefined major mode list when I was testing out some weird config I made for myself. I set global-flycheck-mode to t and off it goes, that's as far as Flycheck config goes for me for 6 years or so. Configuring modern checkers typically is some $ cmd --init or install a few language-specific packages, or in the case of cquery, simple type `$ bear -a make` once. No substantial reading, coding or testing is required.

About memory, are you telling me `(require 'flymake)` doesn't cost memory?

Finally, emailing and subsequently discussing issues on a mailing list is not only obsolete, it's a major time sink due to incessent amount of repeated strong opinions by under-informed people. This is endemic across emacs dev culture.

I thank you a lot for yasnippet, eglot and many other contributions, but I don't fine Flymake the code or documentation particularly valuable, and your attitude this time around certainly didn't help. I don't hold grudges on you, but I surely hope it doesn't happen again.

**Comment by joaotavora** [comment]
> Which part of my original request is impolite?

Indeed nothing was impolite in the original request.  Only later, after I asked some clarifying questions, your answers to these questions became random, gratuitous, reckless and rather *impossible* to decipher.

 And [I've already suggested](https://github.com/joaotavora/eglot/issues/42#issuecomment-406195599) what I think is the best way to get Flycheck working in Eglot.  The suggestion put the conversation on topic, but got 0 comment from Jimmy. Anyway, if someone is familiar with Flycheck's API (I'm not) it should be not much work to adapt an arbitrary Flymake backend, including Eglot's.

> About memory, are you telling me (require 'flymake) doesn't cost memory?

Jimmy, I challenge you to measure the memory impact of `(require 'flymake)` and post the result here.  Then do the same for `(require 'flycheck)` for good measure.

> being defensive when I suggested Flymake to be inferior

Look, if I ask you for something, because your horse and sombrero are clearly inferior and lacking quite a bit, and you ask me *what the hell I'm talking about*, are you being defensive or trying to understand *what the hell I'm talking about*? 

> I'm not going to bother diffing the actual content of the files. 

Who in the good Lord's name ever asked you to diff a file??? 

> Other than having integration with eglot, which part of it is not either me-too or straightly inferior? I don't even want to pull out the comparison table on Flycheck's website

To answer the question, by all means, be my guest and "pull out" the comparison table. I've already I've worked with the developers in that very table: it went very well and civil [with absolutely no disagreements basically because we were all grounded in reality, also know as the truth](https://github.com/flycheck/flycheck/pull/1412) , not in our personal opinions.

Even though it only lists headings for what Flycheck finds relevant, it shows that Flymake is not "lacking quite a bit" by a mile.  

But here are some ways in which Flymake is superior to Flycheck:

* Backends can report multiple times for multiple/visible regions in the buffer, reducing resource comsumption dramatically.  This will allow spell checkers to be written on Flymake, something that Flycheck [notably can't do](https://github.com/flycheck/flycheck/issues/382).

* Multiple backends per buffer can run in parallel. They don't need to be run in series like in Flycheck (yes, [I pulled it out](http://www.flycheck.org/en/latest/user/flycheck-versus-flymake.html#multiple-syntax-checkers-per-buffer));

* Diagnostics can be customized to be interactive buttons, run arbitrary functions, etc.

* Its API doesn't suffer from the serious problems that its mantainer is [having a hard time dealing with](https://lists.gnu.org/archive/html/emacs-devel/2018-07/msg00436.html).

> so it's fine if you don't want to offer any help, as I don't see any willingness so far.

Didn't I point you [THREE](https://github.com/joaotavora/eglot/issues/42#issuecomment-406195244) - [TIMES](https://github.com/joaotavora/eglot/issues/42#issuecomment-406195599) - [ALREADY](https://github.com/joaotavora/eglot/issues/42#issuecomment-406534346) to [the relevant documentation](https://www.gnu.org/software/emacs/manual/html_node/flymake/Extending-Flymake.html#Extending-Flymake) that you stubbornly claim didn't exist???  Geez! What other "pointers" do you want? Do you want me to go there and move your friggin mouse pointer to the link and hand-hold you while you read? 

> repeated strong opinions by under-informed people. This is endemic across emacs dev culture.

Oh, oh, and next you'll argue that this whole exchange shows that you're absolutely innocent of any of that!!

> I don't hold grudges on you, but I surely hope it doesn't happen again.

If that is some kind of threat, it's as hollow as everything else you wrote here.

**Comment by wyuenho** [comment]
As I said, full report comes later when I have time. I have also said that Flymake bothers me enough that I want it gone, but I probably should have added it doesn't bother me enough to want to engage the mailing list, since the whole point of this issue is to find out if there's a quick way to turn Flycheck into an alternative. I'm not obligated to file reports for Flymake, especially when my concern is solely about it being the only choice in eglot OTTB.

The reason I don't even want to reply to your suggestion to "refer the Flycheck developers to this protocol and convince them to create a Flycheck interface for the Flymake backends" is I don't consider that as a genuine answer. If you want Flymake to gain traction again, the proper way to do it is to create bridges for existing Flycheck backends to be plugged into Flymake, not egoistically asking people to ask more people to do it the other way around while being the less capable option. I consider this a fuck-off answer, or an answer concerned solely with your own interests as opposed to the general public. Giving me a couple of sentences of how the eglot flymake backend works, like where data comes from, that the format is etc is what I would consider a sufficent and genuine answer. All I can see now is a few variables and functions that take arbitrary arguments. I don't know what "well-defined" protocol means to you, but this is not "well-defined", this is basically undefined.

I don't know when was the last time you stepped out of your Flymake developer's shoes and start being a user. To you, it seems that Flymake's core features being close to Flycheck does not qualify as "lacking quite a bit", but as a user, my concern is the value its ecosystem provides to me, which isn't much, due to configuration and backend issues, and a few minor UI ones. This experience is vastly worst than Flycheck's by default, so this constitutes as "lacking quite a bit" for me.

About memory, I don't understand what you don't understand. If I'm not being clear all along, I would preferrably see 0 memory usage. I don't care if Flymake uses less or more than Flycheck. It doesn't matter to me unless I can wholesale replace Flycheck with Flymake, which isn't possible unless I limit myself to the couple of languages that Flymake supports.

As to diffing a file, there are only a few ways to confirm what's actually changed when the version number hasn't. One of them is diffing. What's an easier way?

I don't even want to respond to your purported superiority of Flymake. I consider 2 of them irrelevent to eglot, 1 a bug and 1 just totally made up. Maybe I'm not very creative, which is likely to be true, but I just don't see how you can substantially improve upon Flycheck. It also takes some balls to say that 1 person's labor is superior to literally 100s of people's over 6 years, although that's possible.

> If that is some kind of threat, it's as hollow as everything else you wrote here.

This isn't a threat, as I have no effective recourse against your behavior. You keep saying I'm being hollow, and I keep saying that's intentional, as I don't consider the specific details of Flymake's deficiencies relevent to this request nor am I obligated to provide you with the full and complete details of all my objections to it. Flymake's continued existence is valueless, that alone should be sufficient reason, but spelling this out right off the bat doesn't seem very "polite".

Lastly, you seem to have a lot of trouble understanding the human psyche, I sincerely urge you interact with normal humans more.

**Comment by joaotavora** [comment]
>The reason I don't even want to reply to your suggestion to "refer the Flycheck developers to this protocol and convince them to create a Flycheck interface for the Flymake backends" is I don't consider that as a genuine answer.

Not geniune??? You asked me verbatim for  "a few pointers for those of us who prefer flycheck, or better yet, integrate with flycheck in eglot".  This is the best way that I know: find someone familiar with Flycheck's API (presumably its developers), pass them the Flymake API and make an adaptor. 

It's totally fine that you didn't know of the Flymake had an API before, but when I showed it to you, it's like you started freaking out.  Indeed, quite amanzingly, you still to this point maintain that this API does not exist.

> It also takes some balls to say that 1 person's labor is superior to literally 100s of people's over 6 years, although that's possible.

Look, you just prove that you know nothing about software development.  Just because I didn't use their code directly, doesn't mean I don't build on the efforts of those 100s of people!  In fact, every day, directly or indirectly, when building something, I build on the efforts of everyone in the world that has ever lived. Have you ever heard the about "standing on the shoulders of giants"?  I try to stand there proudly with my feet, not my balls.

Stop seeing this like it's a horse race, that's very immature and not at all what software is about, let alone free software. It's not a horse race. And if it were, you would find a horse with 100 people's contribution on top of it is sometimes slower and encumbered than a fresh new one. 

Use Flycheck if you want, be happy.  I never told you not to.  I just talked about Flymake's advantages when you specifically asked me to to. I tried to learn about perceived deficiencies in Flymake, but you ranted on about version numbers and broken 1-var-long customization group for a backend being developed in master.  

> Giving me a couple of sentences of how the eglot flymake backend works, like where data comes from, what the format is etc is what I would consider a sufficent and genuine answer. 

Did you even click the links I gave you a million times already?  What is missing from section "2.2 Backend Functions" in the manual?  Here's how it starts:

```
Flymake backends are Lisp functions placed in the special hook
‘flymake-diagnostic-functions’.

   A backend’s responsibility is to diagnose the contents of a buffer
for problems, registering the problem’s positions, type, and summary
description.  This information is collected in the form of diagnostic
objects created by the function ‘flymake-make-diagnostic’ (*note Flymake
utility functions::), and then handed over to Flymake, which proceeds to
annotate the buffer.
```

Do you want me to go there and read the rest to you in my beautiful singing voice?

> All I can see now is a few variables and functions that take arbitrary arguments. I don't know what "well-defined" protocol means to you, but this is not "well-defined", this is basically undefined.

I realize now that I am basically feeding a troll, but you live in another world.  What are the few variables?  There are only two variables in 26.1, `flymake-diagnostic-functions` and `flymake-diagnostic-types-alist`.  Have you consulted their docstrings?

> As to diffing a file, there are only a few ways to confirm what's actually changed when the version number hasn't. One of them is diffing. What's an easier way?

Read NEWS:

```
*** Flymake has been completely redesigned

Flymake now annotates arbitrary buffer regions, not just lines.  It
supports arbitrary diagnostic types, not just errors and warnings (see
variable 'flymake-diagnostic-types-alist').

It also supports multiple simultaneous backends, meaning that you can
check your buffer from different perspectives (see variable
'flymake-diagnostic-functions').  Backends for Emacs Lisp mode are
provided.

The old Flymake behavior is preserved in the so-called "legacy
backend", which has been updated to benefit from the new UI features.
```

> Lastly, you seem to have a lot of trouble understanding the human psyche, I sincerely urge you interact with normal humans more.

What do you think gives you the right to comment on the kind of humans I interact with? Do I know you in person? Have you no shame?

> I don't even want to respond to your purported superiority of Flymake. I consider 2 of them irrelevent to eglot, 1 a bug and  1 just totally made up

Unless you back this up, especially the accusation of making things up, you are evidencing the psyche that I understand all too well, and that is the psyche of the coward.  Very human indeed, but very sad, too.

**Comment by wyuenho** [comment]
This is how I see it:

1: I ask for something minimal politely about integration with a separate package
2. You ask for why
3. I told you the problems I knew at the moment about that separate package
4. You ask for more clarification about that separate package that I have no interest in using, neither on Emacs 26 or master
5. You told me read some documentation that's super vague documentation about that separate package that I have no interest in using (What exactly are in those `(:key value :key2 value2...)`) and `(:report-key value :report-key2 value2...)`?) , and go convince the dev of the package to support the separate package I have no interest in using
6. I gave you my answer for 4
7. You start being impatient, and told me that I should have read 4 different things that I have no reason of trusting short of reading the source code directly. Then you start ranting about how valuable your time is, as if mine doesn't. Then you just put down an totally unintuitive answer as if I should have known better to look for those unintuitive things in the first place. And then you ask for MORE reasons why I don't want to use that separate package and start making accusations.
8. I gave you a full list of reason why I think that separate package is inferior, and why I said what I said
9. You don't accept those reasons and start behaving like a child, more name calling and accusations are thrown
10. I explain more calmly
11. And then you still don't understand where I'm coming from and start putting words in mouth
12. More cycles of misunderstanding and accusations ensues...

All that it takes a politely point out that I was mistaken in step 5), and that I should be aware of versioning inaccuracies that you already knew about, and then I probably would have responded to you in saying I have no interest in digging further into Flymake immediately as that's not the point of this issue, instead of feeling entitled that I should know everything before opening this question issue? I wouldn't have asked if I knew where to look would I?

Anyway, I'm tired. You have a good weekend.

**Comment by joaotavora** [comment]
I have updated the first answer to the original question to mirror the answer I gave earlier, in case anyone googling for this can find an answer to this otherwise interesting question.

**Comment by nickenchev** [comment]
Can't add much to the conversation except for the fact that flymake doesn't make it easy to see errors without using the mouse, drives me nuts.

**Comment by joaotavora** [comment]
> Can't add much to the conversation 

then what are you doing here?

> except for the fact that flymake doesn't make it easy to see errors without using the mouse

Read up, use `M-x display-local-help` or `C-h .`, and move on.

**Comment by nickenchev** [comment]
> then what are you doing here?

Ok, thanks.

**Comment by Islam0mar** [comment]
[doom/flycheck-eglot.el](https://github.com/hlissner/doom-emacs/blob/develop/modules/tools/lsp/autoload/flycheck-eglot.el)

add the following to `:config`/.emacs

```elisp
  (defun flycheck-eglot-init (checker callback)
    "Clean up errors when done.
CHECKER is the checker (eglot).
CALLBACK is the function that we need to call when we are done, on all the errors."
    (cl-labels
        ((flymake-diag->flycheck-err
          (diag)
          (with-current-buffer (flymake--diag-buffer diag)
            (flycheck-error-new-at-pos
             (flymake--diag-beg diag)
             (pcase (flymake--diag-type diag)
               ('eglot-note 'info)
               ('eglot-warning 'warning)
               ('eglot-error 'error)
               (_ (error "Unknown diagnostic type, %S" diag)))
             (flymake--diag-text diag)
             :end-pos (flymake--diag-end diag)
             :checker checker
             :buffer (current-buffer)
             :filename (buffer-file-name)))))
      ;; NOTE: Setting up eglot to automatically create flycheck errors for the buffer.
      (funcall callback 'finished
               (when eglot--unreported-diagnostics
                 (let ((result nil))
                   (dolist (diag (cdr eglot--unreported-diagnostics))
                     (push (flymake-diag->flycheck-err diag) result))
                   (setq eglot--unreported-diagnostics nil)
                   result)))))

  (defun flycheck-eglot-available-p ()
    (bound-and-true-p eglot--managed-mode))

  (add-hook 'eglot--managed-mode-hook
            (defun eglot-prefer-flycheck-h ()
              (when eglot--managed-mode
                (when-let ((current-checker (flycheck-get-checker-for-buffer)))
                  (unless (equal current-checker 'eglot)
                    (flycheck-add-next-checker 'eglot current-checker)))
                (flycheck-add-mode 'eglot major-mode)
                (when (or (eq major-mode 'c-mode) (eq major-mode 'c++-mode))
                  (flycheck-add-next-checker
                   'eglot
			             '(warning . c/c++-googlelint)))
                (flycheck-mode 1)
                (flymake-mode -1))))

 (flycheck-define-generic-checker 'eglot
    "Report `eglot' diagnostics using `flycheck'."
    :start #'flycheck-eglot-init
    :predicate #'flycheck-eglot-available-p
    :modes '(prog-mode text-mode c++-mode c-mode))

  (push 'eglot flycheck-checkers)

  (add-hook 'eglot--managed-mode-hook #'eglot-prefer-flycheck-h)

```

**Comment by joaotavora** [comment]
> Works well.

Until Eglot decides to change its internal, private implementation details.  Doom has this habit of ignoring `--` variables.  I think the correct solution remains https://github.com/flycheck/flycheck/issues/1592.  Also, what's `c/c++-googlelint`

**Comment by Islam0mar** [comment]
Doom is taking your correct solution. I was tracking the function

**Comment by Islam0mar** [comment]
Doom is taking your correct solution [here](https://github.com/hlissner/doom-emacs/blob/develop/modules/tools/lsp/autoload/flycheck-eglot.el#L26)

I was tracking `eglot-flymake-backend` and decided to discard it. I don't know that I shouldn't use `--` variables :) 

`c/c++-googlelint` see [flycheck-google-cpplint](https://github.com/flycheck/flycheck-google-cpplint). I copied the code from my configurations.

**Comment by joaotavora** [comment]
Good to hear that :-)  Regarding the `--` convention, it's the poor man's `private` or `internal` in Elisp.  See https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html#DOCF26.  

Have fun!

**Comment by zw963** [comment]
> [doom/flycheck-eglot.el](https://github.com/hlissner/doom-emacs/blob/develop/modules/tools/lsp/autoload/flycheck-eglot.el)
> 
> add the following to `:config`/.emacs
> 
> ```emacs-lisp
>   (defun flycheck-eglot-init (checker callback)
>     "Clean up errors when done.
> CHECKER is the checker (eglot).
> CALLBACK is the function that we need to call when we are done, on all the errors."
>     (cl-labels
>         ((flymake-diag->flycheck-err
>           (diag)
>           (with-current-buffer (flymake--diag-buffer diag)
>             (flycheck-error-new-at-pos
>              (flymake--diag-beg diag)
>              (pcase (flymake--diag-type diag)
>                ('eglot-note 'info)
>                ('eglot-warning 'warning)
>                ('eglot-error 'error)
>                (_ (error "Unknown diagnostic type, %S" diag)))
>              (flymake--diag-text diag)
>              :end-pos (flymake--diag-end diag)
>              :checker checker
>              :buffer (current-buffer)
>              :filename (buffer-file-name)))))
>       ;; NOTE: Setting up eglot to automatically create flycheck errors for the buffer.
>       (funcall callback 'finished
>                (when eglot--unreported-diagnostics
>                  (let ((result nil))
>                    (dolist (diag (cdr eglot--unreported-diagnostics))
>                      (push (flymake-diag->flycheck-err diag) result))
>                    (setq eglot--unreported-diagnostics nil)
>                    result)))))
> 
>   (defun flycheck-eglot-available-p ()
>     (bound-and-true-p eglot--managed-mode))
> 
>   (add-hook 'eglot--managed-mode-hook
>             (defun eglot-prefer-flycheck-h ()
>               (when eglot--managed-mode
>                 (when-let ((current-checker (flycheck-get-checker-for-buffer)))
>                   (unless (equal current-checker 'eglot)
>                     (flycheck-add-next-checker 'eglot current-checker)))
>                 (flycheck-add-mode 'eglot major-mode)
>                 (when (or (eq major-mode 'c-mode) (eq major-mode 'c++-mode))
>                   (flycheck-add-next-checker
>                    'eglot
> 			             '(warning . c/c++-googlelint)))
>                 (flycheck-mode 1)
>                 (flymake-mode -1))))
> 
>  (flycheck-define-generic-checker 'eglot
>     "Report `eglot' diagnostics using `flycheck'."
>     :start #'flycheck-eglot-init
>     :predicate #'flycheck-eglot-available-p
>     :modes '(prog-mode text-mode c++-mode c-mode))
> 
>   (push 'eglot flycheck-checkers)
> 
>   (add-hook 'eglot--managed-mode-hook #'eglot-prefer-flycheck-h)
> ```
> 
> Works well.

This answer not works anymore, i create a new gist for this. (sure, stolen from doom-emacs)

https://gist.github.com/zw963/b2f62dccd2b799111a208a2d47791334

**Comment by slackline** [comment]
Perhaps the recent [flycheck-eglot](https://github.com/intramurz/flycheck-eglot) package is relevant to this old thread.
